Update nwb recording extractor to use offsets

h-mayorquin opened the pull request:
The newest version of PyNWB now handles offsets as part of electrical series. This PR modifies the nwb recording extractor so it uses this argument to load offsets.

I also removed the checks for gains == 1. I think this is some remnant of testing because we were adding data to nwbfiles without knowing the exact units.

samuelgarcia commented:
Hi Ramon, thanks for this.
If it depend on newest version of pynwb could your push dependency constrain in releated files:
  * requirements_test.txt
  * requirements_extractor.txt

h-mayorquin commented:
> Hi Ramon, thanks for this. If it depend on newest version of pynwb could your push dependency constrain in releated files:
> 
> * requirements_test.txt
> * requirements_extractor.txt

It was not present in `requirements_test.txt` already. I added it there and updated it on the other file.

alejoe91 commented on spikeinterface/extractors/nwbextractors.py:
```diff
@@ -162,9 +157,15 @@ def __init__(self, file_path: PathType, electrical_series_name: str = None, load
+        # Set gains
+        self.set_channel_gains(gains)
+        # Set offsets
+        self.set_channel_offsets(self._es.offset * 1e6)
```
@h-mayorquin probably we have to check here if `offset` is a field. Otherwise this will make it incompatible with older NWBfiles, correct?

somethink like:
```
if hasattr(self._es, "offset"):
    self.set_channel_offsets(self._es.offset * 1e6)
```

h-mayorquin replied:
I just went and check this with to older files:

https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/src/master/nwb/H19.29.141.11.21.01.nwb
https://dandiarchive.org/dandiset/000003/0.210812.1448/files?location=sub-YutaMouse37%2F (first file)

As long as you are reading with the new pynwb library the attribute will be there (it is zero in all those cases).

alejoe91 replied:
IMO we should check anyways so that the extractor would still be compatible with older `pynwb` versions. What do you think?

CodyCBakerPhD replied:
It's mostly checking against the previous practice of setting offsets as columns in the ElectrodeTable in lieu of having that new offset field in the TimeSeries base class.

So to summarize, if using an older PyNWB version to read an older NWBFile, there should be no offset field in the TimeSeries but it might be in the ElectrodeTable so we should have a condition for reading it from there (though this will be hard to write tests for since it's only possible through older PyNWB versions) just to support older versions (or just say screw it and set a minimum bound on PyNWB version to latest).

If reading an older NWBFile with new PyNWB, there will be an offset field in the TimeSeries but it likely won't be set/read from the ElectrodeTable column where it may have previously been set ad hoc, so there should be a condition there for searching the ElectrodeTable just in case the information is available there to be set for SpikeInterface.

h-mayorquin replied:
@CodyCBakerPhD just one clarification that I am not sure you are aware off (maybe you are). 

In spikeinterface there is a difference between the property "offset" as is stored in the relevant code here:

https://github.com/SpikeInterface/spikeinterface/blob/2fbc9eef8cc4cf7e280f741076b250cbdfcb72ac/spikeinterface/extractors/nwbextractors.py#L199-L203

And the property that is used for the scaling that you get using the `get_channel_offsets` function which is called `offset_to_uv` :

https://github.com/SpikeInterface/spikeinterface/blob/2fbc9eef8cc4cf7e280f741076b250cbdfcb72ac/spikeinterface/core/baserecording.py#L654-L655

So, as thing stand currently the `NwbRecordingExtractor` does not have any offsets in any way. That is, as far as I understand `NwbRecordingExtractor` might read offsets as a property but that property is not used to scale the traces.

h-mayorquin replied:
> IMO we should check anyways so that the extractor would still be compatible with older `pynwb` versions. What do you think?

@alejoe91 
Yes, I guess it makes sense not to break the code if someone can't update to a newer version of `PyNWB`. I implemented this in the latest commit.

h-mayorquin replied:
But as @CodyCBakerPhD was saying maybe we should add functionality so that if the channel offset is either:
1) not found or
2) is zero (as it is when is retrieved with a new version of PyNWB but was not set) 

Then we should check if a property called offset is in the electrode table and try to extract it from there. 

Is that a correct, reading @CodyCBakerPhD ? Would this cover the two cases you mention?

CodyCBakerPhD replied:
> In spikeinterface there is a difference between the property "offset" as is stored in the relevant code here:

Ohh yeah I forgot about that. Thanks for noticing... The 'offset' value attached to the NWB time series should be in Volts, so I guess we should be setting `offset_to_uv = offset_in_nwb * 1e-6`?


> But as @CodyCBakerPhD was saying maybe we should add functionality so that if the channel offset is either:
> not found or
> is zero (as it is when is retrieved with a new version of PyNWB but was not set)
> Then we should check if a property called offset is in the electrode table and try to extract it from there.
> Is that a correct, reading @CodyCBakerPhD ? Would this cover the two cases you mention?

Sounds good

alejoe91 commented on spikeinterface/extractors/nwbextractors.py:
```diff
@@ -162,9 +157,20 @@ def __init__(self, file_path: PathType, electrical_series_name: str = None, load
+        self.set_channel_gains(gains)
+        
+        # Set offsets
+        offset = self._es_offset if hasattr(self._es, "offset") else 0
```
```suggestion
        offset = self._es.offset if hasattr(self._es, "offset") else 0
```
